lib/rjgit.rb

```diff
@@ -355,6 +353,86 @@ def self.successful?(result)
       end
       
     end
+
+    class ApplyPatchToIndex < RJGit::Plumbing::Index
+
+      import org.eclipse.jgit.patch.Patch
+      import org.eclipse.jgit.diff.DiffEntry
+
+      ADD    = org.eclipse.jgit.diff.DiffEntry::ChangeType::ADD
+      COPY   = org.eclipse.jgit.diff.DiffEntry::ChangeType::COPY
+      MODIFY = org.eclipse.jgit.diff.DiffEntry::ChangeType::MODIFY
+      DELETE = org.eclipse.jgit.diff.DiffEntry::ChangeType::DELETE
+      RENAME = org.eclipse.jgit.diff.DiffEntry::ChangeType::RENAME
+
+      def initialize(repository, patch, ref = Constants::HEAD)
+        super(repository)
+        @ref   = ref
+        @patch = Patch.new
+        @patch.parse(ByteArrayInputStream.new(patch.to_java_bytes))
+        raise 'Error parsing patch' unless @patch.getErrors.empty?
+        @current_tree = Commit.find_head(@jrepo, ref).tree
+      end
+
+      def commit(message, author, parents = nil, force = false)
+        super(message, author, parents, @ref, force)
+      end
+
+      def build_map
+        @patch.getFiles.each do |file_header|
+          case file_header.getChangeType
+          when ADD
+            add(file_header.getNewPath, apply('', file_header))
+          when MODIFY
+            add(file_header.getOldPath, apply(getData(file_header.getOldPath), file_header))
+          when DELETE
+            delete(file_header.getOldPath)
+          when RENAME
+            delete(file_header.getOldPath)
+            add(file_header.getNewPath, getData(file_header.getOldPath))
+          when COPY
+            add(file_header.getNewPath, getData(file_header.getOldPath))
+          end
+        end
+      end
+
+      private
+
+      def getData(path)
+        (@current_tree / path).data
+      end
+
+      def hunk_sanity_check(hunk, hunk_line, pos, newLines)
+        raise 'Could not build treemap: the patch failed to apply' unless newLines[hunk.getNewStartLine - 1 + pos] == hunk_line[1..-1]
+      end
+
+      def apply(original, file_header)
```

Comment:
JGit's apply can throw IOExceptions and PatchApplyExceptions, how are you dealing with those in your code?

Reply:
Looking at the Java code, PatchApplyExceptions are thrown in three cases:
* when an IOException is caught.
  * Since this code doesn't involve any direct filesystem IO, I assume we won't encounter this scenario.
* when parsing the patch results in errors.
  * Raising an exception [here](https://github.com/repotag/rjgit/pull/50/files#diff-fc239e234bb85098417c9b398efae7e5R373)
* when a patch fails to apply to the original file.
  * Raising an exception [here](https://github.com/repotag/rjgit/pull/50/files#diff-fc239e234bb85098417c9b398efae7e5R406)

Do you think that in the last two cases, we should raise a PatchApplyExceptions specifically?

Reply:
Now throwing PatchApplyExceptions specifically